nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -975,267 +991,174 @@ ngx_http_modsecurity_init_process(ngx_cycle_t *cycle)
-
-    if (rc != NGX_DECLINED) {
-        return ngx_http_finalize_request(r, rc);
-    }
+    ngx_http_request_t  *r;
 
-    if (ngx_http_modsecurity_save_request_body(r) != NGX_OK
-            || ngx_http_modsecurity_save_headers_in(r) != NGX_OK ) {
+    r = ev->data;
+    r->main->blocked--;
+    r->aio = 0;
 
-        return ngx_http_finalize_request(r, NGX_HTTP_INTERNAL_SERVER_ERROR);
-    }
+    ngx_http_modsecurity_ctx_t *ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
+    ctx->thread_running = 0;
 
-    r->phase_handler++;
     ngx_http_core_run_phases(r);
-    ngx_http_finalize_request(r, NGX_DONE);
 }
 
 
 static ngx_int_t
-ngx_http_modsecurity_header_filter(ngx_http_request_t *r) {
-    ngx_http_modsecurity_loc_conf_t *cf;
-    ngx_http_modsecurity_ctx_t      *ctx;
-    const char                      *location;
-    ngx_table_elt_t                 *h;
-
-    cf = ngx_http_get_module_loc_conf(r, ngx_http_modsecurity);
-    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
-
-    /* already processed, checking redirect action. */
-    if (ctx && ctx->complete 
-               && r->err_status >= NGX_HTTP_MOVED_PERMANENTLY
-               && r->err_status < 308) {
-
-        /* 3XX load redirect location header so that we can do redirect in phase 3,4 */
-        location = apr_table_get(ctx->req->headers_out, "Location");
-
-        if (location == NULL) {
-            return NGX_HTTP_INTERNAL_SERVER_ERROR;
-        }
+ngx_http_modsecurity_task_offload(ngx_http_request_t *r)
+{
+    ngx_http_modsecurity_thread_ctx_t *thread_ctx;
+    ngx_thread_task_t *task;
 
-        h = ngx_list_push(&r->headers_out.headers);
-        if (h == NULL) {
-            return NGX_HTTP_INTERNAL_SERVER_ERROR;
-        }
+    task = ngx_thread_task_alloc(r->pool, sizeof(ngx_http_modsecurity_thread_ctx_t));
+    if (task == NULL) {
+        return NGX_ERROR;
+    }
 
-        h->hash = 1;
-        h->key.data = (u_char *)"Location";
-        h->key.len = ngx_strlen("Location");
-        h->value.data = (u_char *)location;
-        h->value.len = ngx_strlen(location);
+    thread_ctx = task->ctx;
+    thread_ctx->r = r;
 
-        return ngx_http_next_header_filter(r);
-    }
+    task->handler = ngx_http_modsecurity_thread_func;
+    task->event.handler = ngx_http_modsecurity_thread_completion;
+    task->event.data = r;
 
-    if (r != r->main || !cf->enable || ctx == NULL ||ctx->complete) {
-        return ngx_http_next_header_filter(r);
+    ngx_thread_pool_t* thread_pool = ngx_thread_pool_get((ngx_cycle_t *)ngx_cycle, &thread_pool_name);
+    if (ngx_thread_task_post(thread_pool, task) != NGX_OK) {
+        return NGX_ERROR;
     }
 
-    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "modSecurity: header filter");
+    r->main->blocked++;
+    r->aio = 1;
 
-    r->filter_need_in_memory = 1;
     return NGX_OK;
 }
 
 
+static void
+ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
+{
+    r->main->count--;
```

Comment:
what's difference between count and block?

Reply:
I don't know. All modules are doing this, and nginx hangs if I don't.